### ISSUE TITLE
Sequence[Mappin[str, Any]] won't allow dict with different layouts

### ISSUE THREAD

**Issue by gh-andre** [python/mypy#15194]
**Bug Report**

When a function takes `Sequence[Mapping[str, Any]]` and a call is made with dictionaries having different fields, a cryptic error is reported that seems to indicate strange affinity for same type even for different field types.

This use is typical for passing in a Mongo DB pipeline list, which is flagged as an error, while there is none, like in this method:

https://pymongo.readthedocs.io/en/stable/api/pymongo/collection.html#pymongo.collection.Collection.aggregate

**To Reproduce**

```
mypy --strict --check-untyped-defs --strict-equality \
    --warn-redundant-casts --warn-incomplete-stub \
    --warn-unreachable --install-types \
    --non-interactive app.py
```

```python
def fn(p: Sequence[Mapping[str, Any]]) -> None:
    pass
p = [{"x": 1}, {"y": "A"}]
fn(p)                               # <-- error reported
fn([{"x": 1}, {"y": "A"}])          # <-- no error
q = [{"x": "1"}, {"y": "A", "z": "3"}]
fn(q)                               # <-- no error (but will be if any number is unquoted)
```
The error is reported for `p`:

> app.py:6: error: Argument 1 to "fn" has incompatible type "List[object]"; expected "Sequence[Mapping[str, Any]]"  [arg-type]

**Expected Behavior**

`Sequence[Any]` may be used to pass in different types and should allow dictionaries with arbitrarily different layouts for `Sequence[Mapping[str, Any]]`.

**Actual Behavior**

`mypy` enforces some internal logic that cannot be described as enforcing same layout, which is indicated by `q` having very different dictionaries and not triggering this error. However, if any of the numbers in `q` is unquoted, the error is reported for the last `fn` call.

**Your Environment**

- Mypy version used: `mypy 1.2.0 (compiled: yes)`
- Mypy command-line flags: see above
- Mypy configuration options from `mypy.ini` (and other config files): None (not used)
- Python version used: Python 3.10.10

**Comment by hauntsaninja** [comment]
This is expected. In the single line case, mypy is able to use type context from the type of the function that it is calling to figure out the type of the value being passed in. But in the absence of an explicit type annotation, in the multiline case, mypy just has to guess at the type. The unfortunate thing here is the invariance of dict. So mypy sees a list with item 0 of dict[str, str] and item 1 of dict[str, int | str]. Because of invariance, these two types are incompatible, so they're joined to object, rather than something like a dict[str, object] that you may expect.

**Comment by erictraut** [comment]
This is due to mypy's use of join rather than union. If it used unions, it would infer the type of `p` as `list[dict[str, int] | dict[str, str]]` rather than `list[object]`, and the code would type check without any errors (as it does in pyright).

**Comment by gh-andre** [comment]
@hauntsaninja 

Thank you for responding.

> This is expected

It may be expected from the implementation point of view, as you describe, but from the requirement point of view it cannot be expected because the declared type, which is `Sequence[Mapping[str, Any]]` does not impose this constraint, so it's purely implementation getting in the way.

I also noticed that Visual Studio doesn't report this error and from @erictraut's explanation (thank you for jumping in) I can see now why, and from the usability perspective, it makes more sense to consider alternatives here rather than the merged layout.
